server/shadow/shadow_client.c

```diff
@@ -173,73 +251,124 @@ void shadow_client_message_free(wMessage* message)
 	}
 }
 
-BOOL shadow_client_capabilities(freerdp_peer* peer)
+static INLINE void shadow_client_mark_invalid(rdpShadowClient* client, int numRects, const RECTANGLE_16* rects)
 {
-	rdpShadowSubsystem* subsystem;
-	rdpShadowClient* client;
+	int index;
+	RECTANGLE_16 screenRegion;
+	rdpSettings* settings = ((rdpContext*) client)->settings;
 
-	client = (rdpShadowClient*) peer->context;
-	subsystem = client->server->subsystem;
+	EnterCriticalSection(&(client->lock));
 
-	if (subsystem->ClientCapabilities)
+	/* Mark client invalid region. No rectangle means full screen */
+	if (numRects > 0)
 	{
-		return subsystem->ClientCapabilities(subsystem, client);
+		for (index = 0; index < numRects; index++)
+		{
+			region16_union_rect(&(client->invalidRegion), &(client->invalidRegion), &rects[index]);
+		}
 	}
-	return TRUE;
+	else
+	{
+		screenRegion.left = 0;
+		screenRegion.top = 0;
+		screenRegion.right = settings->DesktopWidth;
+		screenRegion.bottom = settings->DesktopHeight;
+
+		region16_union_rect(&(client->invalidRegion),
+		                    &(client->invalidRegion), &screenRegion);
+	}
+
+	LeaveCriticalSection(&(client->lock));
 }
 
-static INLINE void shadow_client_calc_desktop_size(rdpShadowServer* server, int* pWidth, int* pHeight)
+/**
+ * Function description
+ * Recalculate client desktop size and update to rdpSettings
+ *
+ * @return TRUE if width/height changed.
+ */
+static INLINE BOOL shadow_client_recalc_desktop_size(rdpShadowClient* client)
 {
-	RECTANGLE_16 viewport = {0, 0, server->screen->width, server->screen->height};
+	int width, height;
+	rdpShadowServer* server = client->server;
+	rdpSettings* settings = client->context.settings;
+
+	RECTANGLE_16 viewport = {0, 0, server->surface->width, server->surface->height};
 
 	if (server->shareSubRect)
 	{
-		rectangles_intersection(&viewport, &(server->subRect), &viewport); 
+		rectangles_intersection(&viewport, &(server->subRect), &viewport);
+	}
+
+	width = viewport.right - viewport.left;
+	height = viewport.bottom - viewport.top;
+
+	if (settings->DesktopWidth != (UINT32)width || settings->DesktopHeight != (UINT32)height)
+	{
+		settings->DesktopWidth = width;
+		settings->DesktopHeight = height;
+		return TRUE;
 	}
 
-	(*pWidth) = viewport.right - viewport.left;
-	(*pHeight) = viewport.bottom - viewport.top;
+	return FALSE;
 }
 
-BOOL shadow_client_post_connect(freerdp_peer* peer)
+static BOOL shadow_client_capabilities(freerdp_peer* peer)
+{
+	rdpShadowSubsystem* subsystem;
+	rdpShadowClient* client;
+	BOOL ret = TRUE;
+
+	client = (rdpShadowClient*) peer->context;
+	subsystem = client->server->subsystem;
+
+	IFCALLRET(subsystem->ClientCapabilities, ret, subsystem, client);
+	if (!ret)
+		WLog_WARN(TAG, "subsystem->ClientCapabilities failed");
+
+	/* Recalculate desktop size regardless whether previous call fail 
+	 * or not. Make sure we send correct width/height to client */ 
+	(void)shadow_client_recalc_desktop_size(client);
```

Comment:
After CONNECTION_STATE_CAPABILITIES_EXCHANGE stage this callback is invoked. I checked code in libfreerdp/core/peer.c and found the return value of Capabilities doesn't impact the further process. The shadow_client_recalc_desktop_size() is an internal static help function whose return value is '@return TRUE if width/height changed'. In this callback I just use it to ensure the desktop size is correct because the client side WIDTH/HEIGHT is negotiated in CONNECTION_STATE_CAPABILITIES_EXCHANGE stage. So:
1. In any case the shadow_client_recalc_desktop_size need to be called.
2. We don't care whether desktop size is changed in this callback, so the return value of shadow_client_recalc_desktop_size is ignored.